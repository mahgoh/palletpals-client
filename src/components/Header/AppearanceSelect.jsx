import { useTranslation } from 'react-i18next'
import Select from '@/components/Select'
import { useAppearance } from '@/services/appearance'

export default function AppearanceSelect({ className }) {
  const { t } = useTranslation()
  const { appearance, setAppearance, appearances } = useAppearance()

  const options = appearances.map((appearance) => {
    return { value: appearance, label: t(`common.appearance.${appearance}`) }
  })

  const selectedIndex = options.findIndex(
    (option) => option.value === appearance
  )

  const set = (option) => {
    setAppearance(option.value)
  }

  return (
    <Select
      options={options}
      set={set}
      selectedIndex={selectedIndex}
      className={className}
    />
  )
}
