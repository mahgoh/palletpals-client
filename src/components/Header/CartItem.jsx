import { NavLink } from 'react-router-dom'
import { useTranslation } from 'react-i18next'
import { productImageURL, formatPrice } from '@/utils/common'

export default function CartItem({ cartItem }) {
  const { t } = useTranslation()

  const { product, pricePerUnit, quantity } = cartItem

  return (
    <div className="flex py-4 first:pt-0 last:pb-0">
      <div className="mr-4 h-12 w-12 overflow-hidden bg-gray-100 dark:bg-gray-800">
        {product.productImages.length > 0 && (
          <NavLink to={`/product/${product.id}`}>
            <img
              className="h-12 w-12 object-cover"
              src={productImageURL(product.productImages[0].id)}
              alt={product.name}
            />
          </NavLink>
        )}
      </div>
      <div className="grow">
        <NavLink to={`/product/${product.id}`}>
          <h3 className="pb-2 text-lg font-bold leading-10 tracking-tight hover:underline focus:underline sm:text-lg sm:leading-none">
            {product.name}
          </h3>
        </NavLink>
        <div>
          {t('common.amount')}: {quantity}
        </div>
      </div>
      <div className="text-right text-lg font-medium">
        {formatPrice(pricePerUnit * quantity)}
      </div>
    </div>
  )
}
