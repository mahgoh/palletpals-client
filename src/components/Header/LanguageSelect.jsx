import Select from '@/components/Select'
import { useLanguage } from '@/services/language'

export default function LanguageSelect() {
  const { language, setLanguage, languages } = useLanguage()

  const options = languages.map((language) => {
    return { value: language, label: language }
  })

  const selectedIndex = options.findIndex((option) => option.value === language)

  const set = (option) => {
    setLanguage(option.value)
  }

  return <Select options={options} set={set} selectedIndex={selectedIndex} />
}
