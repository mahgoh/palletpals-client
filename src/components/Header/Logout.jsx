import { useNavigate, useLocation } from 'react-router-dom'
import { useTranslation } from 'react-i18next'
import { useAuth } from '@/services/auth'

export default function Logout() {
  const { t } = useTranslation()
  const auth = useAuth()
  const navigate = useNavigate()

  function logout() {
    auth.signout(() => navigate('/'))
  }

  if (auth.user) {
    return (
      <button className="ml-4" onClick={logout}>
        {t('common.auth.logout')}
      </button>
    )
  } else {
    return null
  }
}
