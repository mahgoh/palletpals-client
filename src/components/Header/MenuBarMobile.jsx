import { useState } from 'react'
import { NavLink } from 'react-router-dom'
import PropTypes from 'prop-types'
import { classNames } from '@/utils/common'
import { MenuIcon, XIcon } from '@heroicons/react/outline'
import Spacer from '@/components/Spacer'
import AppearanceSelect from './AppearanceSelect'
import LanguageSelect from './LanguageSelect'

export default function MenuBarMobile({ routes }) {
  const [isOpen, setIsOpen] = useState(false)

  function renderCount(count) {
    if (!count || count === 0) return null

    return (
      <span className="inline-flex h-7 w-7 items-center justify-center rounded-full bg-orange-500/10 text-sm font-semibold text-orange-500">
        {count}
      </span>
    )
  }

  function renderMenu() {
    return (
      <div
        className={classNames(
          'fixed bottom-0 right-0 z-20 h-screen w-full bg-transparent bg-white transition-all duration-300 ease-menu dark:bg-gray-900',
          isOpen
            ? 'bottom-0 scale-100 opacity-100'
            : '-bottom-full scale-150 opacity-0'
        )}
      >
        <div className="mx-auto flex h-20 justify-end px-4">
          <button
            type="button"
            className="px-4"
            onClick={() => setIsOpen(false)}
          >
            <XIcon className="h-8 w-8" />
          </button>
        </div>
        <div className="flex flex-col items-start space-y-2 p-8 text-lg">
          {routes.map((route) => (
            <NavLink
              to={route.to}
              key={route.to}
              className={({ isActive }) => {
                return classNames(
                  'flex w-full items-center justify-between py-2',
                  isActive && 'font-bold text-orange-500'
                )
              }}
              onClick={() => {
                setIsOpen(false)
              }}
            >
              <span>{route.label}</span>
              {renderCount(route?.count)}
            </NavLink>
          ))}
          <Spacer size="sm" />
          <AppearanceSelect className="w-full" />
          <LanguageSelect className="w-full" />
        </div>
      </div>
    )
  }

  return (
    <div className="flex justify-end sm:hidden">
      <button type="button" className="px-4" onClick={() => setIsOpen(true)}>
        <MenuIcon className="h-8 w-8" />
      </button>
      {renderMenu()}
    </div>
  )
}

MenuBarMobile.propTypes = {
  routes: PropTypes.arrayOf(
    PropTypes.exact({
      to: PropTypes.string,
      label: PropTypes.string,
      count: PropTypes.number,
    })
  ).isRequired,
}
