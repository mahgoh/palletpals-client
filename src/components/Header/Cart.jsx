import { useTranslation } from 'react-i18next'
import { ShoppingCartIcon } from '@heroicons/react/outline'
import { useAuth } from '@/services/auth'
import { useCart } from '@/services/cart'
import { LinkButton } from '@/components/Button'
import CartItem from './CartItem'

export default function Cart() {
  const { t } = useTranslation()
  const auth = useAuth()
  const { cart } = useCart()

  function renderCartItems() {
    if (!cart || cart.length === 0) {
      return null
    }

    return cart.shoppingCart.map((cartItem, i) => (
      <CartItem key={i} cartItem={cartItem} />
    ))
  }

  function renderCart() {
    if (!auth.authenticated) {
      return (
        <>
          <div>{t('common.cart.not-authenticated')}</div>
          <LinkButton to="/login" className="mt-4 w-full">
            {t('common.auth.login')}
          </LinkButton>
        </>
      )
    }

    return (
      <>
        <div className="w-full grow justify-start divide-y divide-gray-200 overflow-y-auto dark:divide-gray-700">
          {renderCartItems()}
        </div>
        <LinkButton to="/cart" className="mt-4 w-full">
          {t('common.cart.go')}
        </LinkButton>
      </>
    )
  }

  return (
    <div className="group ml-4 flex items-center">
      <ShoppingCartIcon
        className="h-5 w-5 cursor-pointer text-gray-600 transition-colors hover:text-orange-500 focus:text-orange-500 dark:text-gray-100 dark:hover:text-orange-500"
        aria-hidden="true"
      />
      <div className="absolute top-0 right-0 z-20 hidden h-screen w-screen transition-all group-hover:block md:w-1/3">
        <div className="flex h-full w-full flex-col border-l border-gray-200 bg-white p-8 text-left dark:border-gray-700 dark:bg-gray-900">
          <h2 className="pb-10 pt-4 text-2xl font-extrabold leading-10 tracking-tight sm:text-3xl sm:leading-none">
            {t('common.cart.title')}
          </h2>
          {renderCart()}
        </div>
      </div>
    </div>
  )
}
