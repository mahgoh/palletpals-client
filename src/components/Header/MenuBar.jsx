import PropTypes from 'prop-types'
import Navigation from '@/components/Navigation'
import AppearanceSelect from './AppearanceSelect'
import LanguageSelect from './LanguageSelect'
import Cart from './Cart'
import Logout from './Logout'

export default function MenuBar({ routes }) {
  return (
    <div className="hidden justify-end sm:flex">
      <Navigation routes={routes} />
      <Cart />
      <div className="ml-4 flex items-center space-x-2">
        <AppearanceSelect className="w-28" />
        <LanguageSelect />
      </div>
    </div>
  )
}

MenuBar.propTypes = {
  routes: PropTypes.arrayOf(
    PropTypes.exact({
      to: PropTypes.string,
      label: PropTypes.string,
    })
  ).isRequired,
}
