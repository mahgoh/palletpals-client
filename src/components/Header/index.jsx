import { Link } from 'react-router-dom'
import { useTranslation } from 'react-i18next'
import { useAuth } from '@/services/auth'
import { useCart } from '@/services/cart'
import MenuBar from './MenuBar'
import MenuBarMobile from './MenuBarMobile'

export default function Header() {
  const { t } = useTranslation()
  const { authenticated, isAdmin } = useAuth()
  const { cart } = useCart()

  const routes = {
    home: {
      to: '/',
      label: t('common.home'),
    },
    admin: {
      to: '/admin',
      label: t('common.admin'),
    },
    products: {
      to: '/products',
      label: t('common.product.title', { numProducts: 2 }),
    },
    profile: {
      to: '/profile',
      label: t('common.profile.title'),
    },
    cart: {
      to: '/cart',
      label: t('common.cart.title'),
      count: cart.shoppingCart.length,
    },
    login: {
      to: '/login',
      label: t('common.auth.login'),
    },
    logout: {
      to: '/logout',
      label: t('common.auth.logout'),
    },
  }

  let routesMain = [routes.products, routes.login]
  let routesMobile = [routes.home, routes.products, routes.login]

  if (authenticated) {
    routesMain = [routes.products, routes.profile, routes.logout]

    routesMobile = [
      routes.home,
      routes.products,
      routes.profile,
      routes.cart,
      routes.logout,
    ]

    if (isAdmin) {
      routesMain.unshift(routes.admin)
      routesMobile.splice(1, 0, routes.admin)
    }
  }

  return (
    <header className="sticky top-0 z-20 h-20 border-b border-gray-200 bg-white dark:border-gray-700 dark:bg-gray-900">
      <div className="mx-auto grid h-full grid-cols-2 px-4 lg:max-w-6xl lg:px-10">
        <Link
          to="/"
          className="inline-flex items-center text-3xl font-extrabold tracking-tight"
        >
          <span className="bg-gradient-to-r from-orange-500 to-rose-500 bg-clip-text text-transparent">
            PalletPals
          </span>
        </Link>
        <MenuBar routes={routesMain} />
        <MenuBarMobile routes={routesMobile} />
      </div>
    </header>
  )
}
