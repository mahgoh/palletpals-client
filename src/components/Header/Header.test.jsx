import React from 'react'
import { render, screen } from '@/utils/test-utils'
import Header from '.'

describe('header', () => {
  it('should render', () => {
    const { container } = render(<Header />, { renderer: 'complete' })

    expect(screen).not.toBeFalsy()
    expect(container.querySelector('header')).toMatchSnapshot()
  })
})
