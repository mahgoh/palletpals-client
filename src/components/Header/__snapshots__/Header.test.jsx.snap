// Vitest Snapshot v1

exports[`header > should render 1`] = `
<header
  class="sticky top-0 z-10 h-20 border-b border-gray-200 bg-white dark:border-gray-700 dark:bg-gray-900"
>
  <div
    class="mx-auto grid h-full grid-cols-2 px-4 lg:max-w-6xl lg:px-10"
  >
    <a
      class="inline-flex items-center text-3xl font-extrabold tracking-tight"
      href="/"
    >
      <span
        class="bg-gradient-to-r from-orange-500 to-rose-500 bg-clip-text text-transparent"
      >
        PalletPals
      </span>
    </a>
    <div
      class="flex justify-end"
    >
      <nav
        class="flex h-full items-center justify-end"
      >
        <a
          class="inline-flex items-center pr-6 last:pr-0"
          href="/first"
        >
          First
        </a>
        <a
          class="inline-flex items-center pr-6 last:pr-0"
          href="/second"
        >
          Second
        </a>
      </nav>
      <div
        class="ml-4 flex items-center space-x-2"
      >
        <div
          class="relative w-28"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="relative w-full cursor-default border border-transparent bg-white py-2 pl-3 pr-10 text-left hover:border-gray-300 focus:border-orange-500 focus:outline-none focus:ring-1 focus:ring-orange-500 dark:bg-gray-900 dark:hover:border-gray-700 sm:text-sm"
            id="headlessui-listbox-button-1"
            type="button"
          >
            <span
              class="block truncate uppercase"
            >
              System
            </span>
            <span
              class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5 text-gray-400"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10 3a1 1 0 01.707.293l3 3a1 1 0 01-1.414 1.414L10 5.414 7.707 7.707a1 1 0 01-1.414-1.414l3-3A1 1 0 0110 3zm-3.707 9.293a1 1 0 011.414 0L10 14.586l2.293-2.293a1 1 0 011.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="relative"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="relative w-full cursor-default border border-transparent bg-white py-2 pl-3 pr-10 text-left hover:border-gray-300 focus:border-orange-500 focus:outline-none focus:ring-1 focus:ring-orange-500 dark:bg-gray-900 dark:hover:border-gray-700 sm:text-sm"
            id="headlessui-listbox-button-2"
            type="button"
          >
            <span
              class="block truncate uppercase"
            >
              en
            </span>
            <span
              class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2"
            >
              <svg
                aria-hidden="true"
                class="h-5 w-5 text-gray-400"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M10 3a1 1 0 01.707.293l3 3a1 1 0 01-1.414 1.414L10 5.414 7.707 7.707a1 1 0 01-1.414-1.414l3-3A1 1 0 0110 3zm-3.707 9.293a1 1 0 011.414 0L10 14.586l2.293-2.293a1 1 0 011.414 1.414l-3 3a1 1 0 01-1.414 0l-3-3a1 1 0 010-1.414z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</header>
`;
